### ISSUE TITLE
How could we install it?

### ISSUE THREAD

**Issue by Montoyez** [alfarok/DynamoViewport#15]
Hi Keith,

Congratulations for creating this fantastic functionality. I heart about it in the Hackathon of London 2019 and I wanted to test it, It looks really cool in Youtube to play with the geometry while orbiting. I have installed Revit 2020 and Dynamo 2.2.1.5291, together with Dynamo Core 2.2.1.5175 and also DynamoSandbox 2.0.2 and DynamoSandbox 2.4.0. 
I followed your steps to clone the Repo and paste it in the different package folders, but nothing is showing when running Dynamo in any version. 
If I could be put in the right direction, it will be really appreciated. 

Thanks in advance.
Bests,
Sergio

**Comment by alfarok** [comment]
Hey Sergio, thanks for checking out the project.  I haven't made any changes in a while so it could be that a few additional installation steps are now required.

1) When you attempt to build the project in visual studio does it build successfully or are there failures?  
2) Do the package files get copied as expected during the build events?

Your `Viewport` dynamo package folder should look like this:
- bin
  - ViewportViewExtension.dll
  - Viewport_ViewExtensionDefinition.xml
 - extra
   - assests
   - index.html
   - styles.css
- pkg.json

And this package should be copied into the appropriate packages folder such as this in Revit 2020:
`C:\..\Roaming\Dynamo\Dynamo Revit\2.3\packages`

I gave it a quick try and it seems to be working but requires the following 2 code changes due to Dynamo making an API breaking change in [pull request #9691](https://github.com/DynamoDS/Dynamo/commit/b8b80a1c6f1fb7e031853c09759350327ef1acf2#diff-ade9eb3fb1e8cf7ec0e008cbb7a7575dL1653):

![image](https://user-images.githubusercontent.com/13341935/80542789-acb8db00-897b-11ea-8981-906503f22a50.png)

[this line](https://github.com/alfarok/DynamoViewport/blob/master/src/ViewportViewExtension/ViewportViewExtension/ViewportViewController.cs#L118) and [this line](https://github.com/alfarok/DynamoViewport/blob/master/src/ViewportViewExtension/ViewportViewExtension/ViewportViewController.cs#L169) need to be updated to 
```C#
var libraryViewColumn = dynamoWindow.FindName("LeftExtensionsViewColumn") as ColumnDefinition;
```

`LibraryViewColumn` was renamed to `LeftExtensionsViewColumn` when `RightExtensionsViewColumn` was introduced with the new "extensions panel" was added I believe.  I am not familiar with the new API but you can probably use the right panel instead of replacing the library on the left now.

Feel free to make a new pull request with any of your updates, cheers!

**Comment by alfarok** [comment]
I made the mentioned [changes](https://github.com/alfarok/DynamoViewport/commit/2123fd372ee24fbd924c96b0785e57df0f049e33) from above if you want to pull the code again.

I also removed this xcopy step from the post build steps:
```
xcopy "$(ProjectDir)DynamoPackage" "$(AppData)\Dynamo\Dynamo Core\2.2\packages" /Y /E
```
It attempts to copy the final `Viewport` package folder to the Dynamo Core packages location for version 2.2.  You can re-add it if you wish to overwrite the package every time you build in visual studio.  Make sure to update the version number accordingly, 2.2 to 2.3, etc.

**Comment by mjkkirschner** [comment]
@alfarok - maybe use the new sidebar API instead to add it to the right side? I think that exists in 2.3... but not sure - 2.5 is released now though.
https://github.com/DynamoDS/Dynamo/blob/54eaefff977fa8f597a589cbf6186ef7cc933823/src/DynamoCoreWpf/Extensions/ViewLoadedParams.cs#L99

**Comment by alfarok** [comment]
Hey @Montoyez, let us know if you are able to get it running.  I would use the approach mentioned above by @mjkkirschner as this is the proper way to load and dock view extensions in newer versions of Dynamo.  At the time that the project was created these APIs did not yet exist.

**Comment by Montoyez** [comment]
Hi @alfarok,

Thanks so much for the quick response and for adding all those improvements to the code, I can´t wait to make it work. Unfortunately I did not manage to see the content inside the package folder, I have to say that I am quite new in coding. 
These are the steps that I have done:

- I have created a new folder here called DynamoViewport:
- C:\Program Files (x86)\DynamoViewport
- I have gone to properties in that folder and allow all allow Security access to all users. 
- With the desktop Github app, I have clone your path to this folder, so now all those files are there.
- Then in Visual Studio I have created a new project pointing to this folder, then in the Solution Explorer I have done right click in ViewportViewExtension.sln file and Build. All looks like fine unless one Error, which says:

Description:
This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is ..\packages\cef.redist.x64.3.3325.1758\build\cef.redist.x64.props.	

Project:
C:\Program Files (x86)\DynamoViewport\src\ViewportViewExtension\ViewportViewExtension.sln	

File:
C:\Program Files (x86)\DynamoViewport\src\ViewportViewExtension\ViewportViewExtension\ViewportViewExtension.csproj	222	

I have been reading about NuGet and tried to update those packages but no success, I have no idea about what they are and how to Restore them properly. I am sure it is a small thing what I am missing but I am definitely blocked. 

Any further advice would be awesome. 

Bests,
Sergio M

**Comment by alfarok** [comment]
Hey Sergio,

I think the issue right now is with building the project.  Everything seems okay up until after you clone the repository.  I pushed a couple other changes that should make it easier so please re-pull the latest changes before proceeding.  Instead of right-clicking and trying to build from the context menu try and open the `.sln` file using visual studio.

Once in visual studio try restoring the NuGet dependencies like this:

![image](https://user-images.githubusercontent.com/13341935/80820056-2c28f300-8ba4-11ea-8da6-9c3aa5f5ca6b.png)

Go to

- Tools -> NuGet Package Manager

- Click `restore` on the top flag that says `Some NuGet packages are missing from this solution...`

- Now try building with the dependencies restored

![image](https://user-images.githubusercontent.com/13341935/80820234-83c75e80-8ba4-11ea-8ff3-9f9ac19a09c4.png)

If successful you should see `DynamoViewport\src\ViewportViewExtension\ViewportViewExtension\DynamoPackage\Viewport` populated with the following items:

![image](https://user-images.githubusercontent.com/13341935/80821431-b3776600-8ba6-11ea-8820-436560964927.png)

Now you can manually copy this `Viewport` folder (inside `DynamoPackage`) to your Dynamo packages folder such as `C:\..\Roaming\Dynamo\Dynamo Revit\2.3\packages` or add a post-build step to automatically copy upon building like this:

```
xcopy "$(ProjectDir)DynamoPackage" "$(AppData)\Dynamo\Dynamo Core\2.3\packages\" /Y /E
```

![image](https://user-images.githubusercontent.com/13341935/80821661-2a146380-8ba7-11ea-913c-26e1486a03f1.png)

OR

un-comment [these lines](https://github.com/alfarok/DynamoViewport/blob/master/src/ViewportViewExtension/ViewportViewExtension/ViewportViewExtension.csproj#L219), and reload the project + rebuild in visual studio instead.

Let me know if this unblocks you and thanks for encouraging me to update this project to keep it at least loading in newer versions of Dynamo!

**Comment by Montoyez** [comment]
Hi @Alfarok,

Thanks for the replay and for the detailed explanation. I have re do all the process from scratch following that workflow and unfortunately I get even more errors. Please find them pasted here. 

![2020-05-02_01-51-49](https://user-images.githubusercontent.com/23310681/80849518-c4ac8b00-8c17-11ea-9bae-240d3e77455b.jpg)

![2020-05-02_01-53-25](https://user-images.githubusercontent.com/23310681/80849524-c9713f00-8c17-11ea-84c8-a154ea559083.jpg)

Any idea about them?

Thanks in advance. 
Bests,
Sergio

**Comment by alfarok** [comment]
What is the build architecture set to in visual studio?

![image](https://user-images.githubusercontent.com/13341935/80850215-c27f0800-8be8-11ea-9319-de3665a8782a.png)

It should be set to `x64`, I would guess that it's set to `Any CPU` based on the error message.  Try switching it and rebuilding.

**Comment by Montoyez** [comment]
Hi @alfarok, 

Thanks for that, I think we are almost there. This time I only got one Warning.
![image](https://user-images.githubusercontent.com/23310681/80861984-066a1f80-8c72-11ea-95b7-b951c978b2e8.png)
Any idea?

Thank you  VERY much. 
SM

**Comment by alfarok** [comment]
It looks like it built successfully!  Although it's something that should be addressed in this case, warnings usually won't prevent the build from succeeding.  You can view the build output window in Visual Studio using `View > Output`, or press`Ctrl+Alt+O` to see the build status.  Also check the contents of this folder: 

`DynamoViewport\src\ViewportViewExtension\ViewportViewExtension\DynamoPackage\Viewport`

If everything listed in the first response is there you should be good to go.

**Comment by Montoyez** [comment]
Hi @alfarok,

Yes, thanks for that. The Built was finally perfectly done but I was looking to the wrong DynamoPackage folder. So now I pasted it to the folder where I have my Dynamo packages and I finally can see it in View tab in Dynamo. The only thing is that the program crashes when I try to run it. 

![image](https://user-images.githubusercontent.com/23310681/80870313-ef452500-8ca5-11ea-9942-a1c8995a672c.png)

- This is what the error says. 

Object reference not set to an instance of an object.

   Object reference not set to an instance of an object.

   at ViewportViewExtension.ViewportViewController.AddViewportView()
   at ViewportViewExtension.ViewportViewExtension.<>c__DisplayClass7_0.<Loaded>b__0(Object sender, RoutedEventArgs args)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.Controls.MenuItem.InvokeClickAfterRender(Object arg)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)


Is it maybe my laptop not powerful enough?

Bests,

Sergio

**Comment by alfarok** [comment]
What does your Dynamo packages directory look like and what are the contents of your `Viewport` package folder?  

Are there any duplicate copies of the package anywhere from your previous work?  

Something is going wrong when the package in instantiating which is usually related to incorrect or duplicate package artifacts.

To debug locally and get the exact error you can also start Revit and attach it as a process in Visual Studio under `Debug -> Attach to Process`, select `Revit.exe` from the list, and hit `Attach`. From this point launch Dynamo and try to launch the extension.  An error should be hit in Visual Studio as the crash occurs.

**Comment by Montoyez** [comment]
Hi @alfarok,

My Dynamo Package Directory looks like this:
![image](https://user-images.githubusercontent.com/23310681/80917897-15c89600-8d62-11ea-9f2e-45b119e70218.png)


The content of my Viewport package folder are:
![image](https://user-images.githubusercontent.com/23310681/80917873-f6316d80-8d61-11ea-8927-fb1a96884f92.png)

And I have installed these other Dynamo Versions for Revit, which they have not the Viewport folder inside their Packages folder:

![image](https://user-images.githubusercontent.com/23310681/80917950-7a83f080-8d62-11ea-984c-447df45cab18.png)

Finally I have attached to process selecting the Revit.exe. This is what I see in VS:

![image](https://user-images.githubusercontent.com/23310681/80918854-4fe86680-8d67-11ea-8b58-e722ec5f0faf.png)
![image](https://user-images.githubusercontent.com/23310681/80918860-54ad1a80-8d67-11ea-862e-ff33c479cba3.png)

![image](https://user-images.githubusercontent.com/23310681/80918888-75757000-8d67-11ea-9b50-7b4f007c0c03.png)

Any suggestion?

Thanks so much,

SM

**Comment by alfarok** [comment]
Does the same thing happen if you delete `Viewport Fun`?

**Comment by Montoyez** [comment]
Hi @alfarok! 

Finally worked out!! :D How excited I am!! Thank you very very much for the massive support and all those tips. I also learnt a lot about VS which made me so motivated to continue. So finally it worked perfectly when I run it with DynamoSandbox - 2.4.0.6186.
With Revit 2020 I and Dynamo 2.2 I always crashed, even removing all Viewport Fun or leaving empty other Dynamo versions Package folders. 

Please have a look to the video that I have done. 

![Dynamo Viewport 1 GIF](https://user-images.githubusercontent.com/23310681/80921257-98a71c00-8d75-11ea-9953-ef056fac444f.gif)

![Dynamo Viewport 2 GIF](https://user-images.githubusercontent.com/23310681/80921258-9b097600-8d75-11ea-947f-a7e6739f79d3.gif)


I can´t wait to create more videos like this with more advanced geometry that I am used to create. Also now I am trying to figure out how to integrate this 3d viewer in a Web, as I reed something. If people could interact with the geometry when visiting a web, would be really cool. If you have any idea, feel free to share.

Finally find here screen shots of Visual Studio after adding  Revit.exe as an attach process. I would love to make it work with Revit Dynamo as it has more functionalities but I also heard that Dynamo core is more powerful so maybe my laptop is not strong enough. 

![image](https://user-images.githubusercontent.com/23310681/80921378-4e726a80-8d76-11ea-9ba4-8bb1b3056c40.png)

![image](https://user-images.githubusercontent.com/23310681/80921380-529e8800-8d76-11ea-96d0-242de2510994.png)


Thanks a lot! This is so cool.

Bests,
Sergio

**Comment by alfarok** [comment]
Glad you got it working!  In the current state the project is just a prototype completed over a couple weeks so feel free to make a pull request in the future with any enhancements you see fit.  

In regards to your Revit issue, I am using Dynamo 2.3 with Revit 2020. I see that you are still using 2.2.  The [API change](https://github.com/DynamoDS/Dynamo/commit/b8b80a1c6f1fb7e031853c09759350327ef1acf2#diff-dda9b3c01263fc46c89f640ddd2c26f5L199) we talked about above occurred between versions 2.2 and 2.3.  Try updating your Dynamo version to 2.3.  Your issue is because 2.2 is now looking for the 2.3 naming.  This is also why it works in Sandbox 2.4 in your case.

I am going to close this issue since it seems to be working but feel free to open an additional one if something else comes up.

Cheers!
